db/backup/recovery_test.py

```diff
@@ -0,0 +1,246 @@
+""" A script to test that tests the validity of a database backup
+in the event of recovery.
+NOTE: DO NOT RUN THIS SCRIPT ON A PRODUCTION SERVER.
+It will drop the local mysql database after each run.
+
+"""
+
+import os
+import unittest
+import subprocess
+from collections import namedtuple
+from google.cloud import storage
+import mysql.connector
+
+
+BACKUP_BUCKET = 'cpg-sm-backups'
+LOCAL_BACKUP_FOLDER = 'latest_backup'
+DATABASE = 'sm_production'
+
+ConnectionDetails = namedtuple('ConnectionDetails', 'address user')
+
+
+class TestDatabaseBackup(unittest.TestCase):
+    """ Testing validity of DB backup"""
+
+    @classmethod
+    def setUpClass(cls):
+        """ Pull the backup file, and restore the database."""
+        backup_folder = pull_latest_backup()
+        get_timestamp(backup_folder)
+        # Pull the latest backup
+        subprocess.run(['mkdir', LOCAL_BACKUP_FOLDER], check=True)
+        subprocess.run(
+            [
+                'gsutil',
+                '-m',
+                'cp',
+                '-r',
+                f'gs://cpg-sm-backups/{backup_folder}',
+                LOCAL_BACKUP_FOLDER,
+            ],
+            check=True,
+        )
+        # Prepare the backup for restoration
+        subprocess.run(
+            [
+                'mariabackup',
+                '--prepare',
+                f'--target-dir={LOCAL_BACKUP_FOLDER}/{backup_folder}',
+            ],
+            check=True,
+        )
+        # Stop the MariaDB server
+        subprocess.run(['sudo', 'systemctl', 'stop', 'mariadb'], check=True)
+
+        if os.path.isdir('/var/lib/mysql/'):
+            print('Backup cannot be performed unless /var/lib/mysql is empty.')
+        else:
+            print('Performing backup')
+        # Perform backup
+        subprocess.run(
+            [
+                'sudo',
+                'mariabackup',
+                '--copy-back',
+                f'--target-dir={LOCAL_BACKUP_FOLDER}/{backup_folder}',
+            ],
+            check=True,
+        )
+        # Fix permissions
+        subprocess.run(
+            ['sudo', 'chown', '-R', 'mysql:mysql', '/var/lib'],
+            check=True,
+        )
+        # Start the MariaDB server
+        subprocess.run(['sudo', 'systemctl', 'start', 'mariadb'], check=True)
+
+    def setUp(self):
+        self.prod_connection = ConnectionDetails('10.152.0.2', 'backup')
```

Comment:
For the prod connection:

- We can't guarantee the internal IP won't change, so we should use internal url.
- Aren't there passwords that go with this? Can we put most of these into environment variables (or a secret) so they can be changed and not captured with the script?

(Might need to add another comment to the script on how to set this up / run it).

Reply:
Great point regarding the internal IP -- I've switched this out to the URL. I didn't originally set passwords, because I instead set the permissions for the new accounts to only work at certain IP addresses -- but in hindsight, that's not a great idea. I've created passwords and moved all this information including into a secret.